Halt unresponsive database requests

marciobarbosa opened the pull request:
Currently, in certain situations, requests to the database may hang
indefinitely. This undesirable behavior can happen when some specific
packets are dropped during the communication between the requestor and
the database. As a result, jobs waiting for the response of the request
in question get stuck in the same state and have to be restarted
manually.

To avoid this problem, add a new configuration directive (timeout) so
the user can specify the maximum number of seconds a database request
can take before being interrupted.

adeason commented on lib/AFS/CellCC/Config.pm:
```diff
@@ -79,12 +79,18 @@ my @directives = (
     { key => 'db/rw/user', },
     { key => 'db/rw/pass', },
 
+    { key => 'db/rw/timeout', type => 'HASH', default => {}, },
```
this should default to the 'ro' value, like we do with the dsn a few lines up. I wouldn't worry about trying to "merge" the sub-values together or whatever; just if the hash isn't specified at all, we should default to the 'ro' values.

marciobarbosa replied:
Ack.

adeason commented on lib/AFS/CellCC/DB.pm:
```diff
@@ -69,19 +69,35 @@ _connect($$$$) {
+    my %timeout = %{ config_get('db/rw/timeout') };
+
+    while (my ($key, $value) = each %timeout) {
+	$opts{$key} = $value;
```
okay, so this is setting any arbitrary dbi option. that's useful, but then it shouldn't be called "timeout", since it's used for anything. just calling it "options" I think sounds fine?

marciobarbosa replied:
Ack.

adeason reviewed (changes requested):
> Halt unresponsive database requests

[I can't seem to make an inline comment on the commit message? oh well]

I suggest calling this something like "Specify timeout for database requests".

in addition to what's inline:

ideally we'd have some kind of general "timeout" parameter that we could just give to dbi, and the engine figures out the relevant db-specific options to set. but it looks like that doesn't exist? so this way of specifying options looks good, but it really is a general-purpose "database option" directive, not a timeout directive.

for mysql, I think we should set a default timeout, since (if I understand correctly), the default behavior allows us to hang literally forever. ideally maybe we'd have some db-agnostic behavior to interrupt db ops if they take too long, but the solutions along those lines I immediately see look ugly (using SIGALRM and such).

we could set a mysql-specific default by checking DBI::parse_dsn() (or a similar DBIx::Simple call if one exists), and just setting the default options if we have mysql (or whatever the driver is called). the default can be pretty high (several minutes, at least); I just want it to not hang forever.

or would mysql-specific directives be ignored if there's another driver? if you could test that, that would be easier. that is, just always set mysql_connect_timeout et al, and see if using another driver just ignores the values, or if it throws an error. if they just get ignored, I'd just do that. (just try a simple test connecting to sqlite or something)

marciobarbosa committed:
Specify timeout for database requests

Currently, in certain situations, requests to the database may hang
indefinitely, at least when using mysql. This undesirable behavior can
happen when some specific packets are dropped during the communication
between the requestor and the database. As a result, jobs waiting for
the response of the request in question get stuck in the same state
and have to be restarted manually.

To avoid this problem, add a new configuration directive (options) so
the admin can add / alter the settings to be used when contacting the
database, including the maximum number of seconds a database request
can take before being interrupted. If not specified through this new
directive, set the default timeout to 5 minutes.

marciobarbosa commented:
> I suggest calling this something like "Specify timeout for database requests”.

Ack.

> ideally we'd have some kind of general "timeout" parameter that we could just
> give to dbi, and the engine figures out the relevant db-specific options to set.
> but it looks like that doesn't exist?

As far as I can tell, no.

> ideally maybe we'd have some db-agnostic behavior to interrupt db ops if they
> take too long, but the solutions along those lines I immediately see look ugly
> (using SIGALRM and such).

Yes, the traditional way to implement a timeout is using $SIG{ALRM}.
https://metacpan.org/pod/distribution/DBI/DBI.pm#Signal-Handling-and-Canceling-Operations

> or would mysql-specific directives be ignored if there's another driver?

Yes, they are ignored.

> if they just get ignored, I'd just do that.

Ack.

marciobarbosa commented:
Updated.

adeason commented on lib/AFS/CellCC/DB.pm:
```diff
@@ -66,22 +66,54 @@ _connect($$$$) {
+    # following types of requests to timeout after a given number of seconds.
+    # If not specified by the admin, set the default timeout to 5 minutes. Note
+    # that mysql-specific directives are ignored by other drivers.
+    if (!exists($opts->{'mysql_connect_timeout'})) {
```
to avoid some repeition here, how about:

for my $option ('mysql_connection_timeout', 'mysql_read_timeout', 'mysql_write_timeout') {
    if (!exists($opts->{$option})) {
        $opts->{$option} = $timeout;
    }
}

I believe this could also be shorter as:

$opts->{$option} //= $timeout;

but I think that requires some version of perl that's after our stated requirement of 5.08. and it's less readable; no real reason to do that, I guess.

adeason commented on doc/pod1/cellcc_config.pod:
```diff
@@ -103,12 +103,40 @@ mysql database called "cellcc" on host db.example.com using port 3306.
+=item B<db/ro/options>
+
+This optional directive can be used to add / alter the settings to be used when
+contacting the database. For example, the admin can specify the timeout to use
```
the current convention in the manpage is to say "you" instead of "the admin". I think talking directly to the user is fine

adeason commented on doc/pod1/cellcc_config.pod:
```diff
@@ -103,12 +103,40 @@ mysql database called "cellcc" on host db.example.com using port 3306.
+This optional directive can be used to add / alter the settings to be used when
+contacting the database. For example, the admin can specify the timeout to use
+when reading or connecting to the B<db/ro/dsn> database. For the MySQL driver,
+the following timeout options are available:
```
any DBI option is fine; we don't need to list these. but we can/should say what our default is, which is these three directives set to 5 minutes

adeason commented on doc/pod1/cellcc_config.pod:
```diff
@@ -103,12 +103,40 @@ mysql database called "cellcc" on host db.example.com using port 3306.
+- mysql_read_timeout
+- mysql_write_timeout
+
+In the following example, requests to the server will timeout after 5 seconds:
```
specifically, this sets the timeout for connecting and for read requests to 5 seconds, but leaves write requests at the default timeout of 5 minutes, yes? we don't need to go into the nuances of the different timeout options, but we should at least be correct here (or just use an easier-to-explain example while still being correct)

adeason reviewed (changes requested):
code looks correct at this point; comments are about minor code changes and docs, etc. I may push suggested edits in a little bit (if I'm remembering correctly that github lets me do that)

adeason commented:
okay, I think I pushed that correctly, but this is not tested or run. could you try running this? (and of course review, if this looks okay to you)

marciobarbosa commented:
> okay, I think I pushed that correctly, but this is not tested or run. could you try running this?
> (and of course review, if this looks okay to you)

Yes, I can do that. Thanks.

marciobarbosa commented:
Passed all tests and looks good to me. +1.
